llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -2764,6 +2766,33 @@ static Constant *ConstantFoldIntrinsicCall2(Intrinsic::ID IntrinsicID, Type *Ty,
                   ? *C0
                   : *C1);
 
+    case Intrinsic::scmp:
+    case Intrinsic::ucmp:
+      if (isa<PoisonValue>(Operands[0]) || isa<PoisonValue>(Operands[1]))
+        return PoisonValue::get(Ty);
+
+      if (!C0 || !C1)
+        return UndefValue::get(Ty);
+
+      if (ICmpInst::compare(*C0, *C1,
+                            IntrinsicID == Intrinsic::scmp
+                                ? ICmpInst::ICMP_SLT
+                                : ICmpInst::ICMP_ULT)) {
+        return ConstantInt::get(Ty, -1, true);
+      }
+      if (ICmpInst::compare(*C0, *C1, ICmpInst::ICMP_EQ)) {
+        return ConstantInt::get(Ty, 0);
+      }
+      if (ICmpInst::compare(*C0, *C1,
+                            IntrinsicID == Intrinsic::scmp
+                                ? ICmpInst::ICMP_SGT
+                                : ICmpInst::ICMP_UGT)) {
+        return ConstantInt::get(Ty, 1);
+      }
+
+      assert(false && "Integer values must compare as equal, or one must be "
```

Comment:
assert false is a bit odd. You could try `llvm_unreachable("It should be unreachable")` instead.